### PR TITLE
feat: add eth_getStorageAt

### LINKED ISSUE #45: eth_api: Implement `get_storage` method
### **Description:**
We need to implement the [`get_storage`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L1042C8-L1042C19) method, which will be responsible for fetching the storage value from a position in a contract.

```rust
/// Retrieves the storage value from a position in a contract.
fn get_storage(&self, address: H160, position: U256, block: Option<BlockIdentifier>) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<Option<H256>>> {
    not_implemented("get_storage")
}
```

#### Expected response
Command:
```
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_getStorageAt",
       "params": ["0x123456789abcdef123456789abcdef1234567890", "0x0", "latest"]
     }'
```
Response:
```
{"jsonrpc":"2.0","result":"0x0000000000000000000000000000000000000000000000000000000000000000","id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - This method should take in three parameters: an `H160` address, a `U256` position, and an optional `BlockIdentifier` to specify the block from which the storage should be retrieved.
    - It should return an `Option<H256>` value representing the storage at the specified position.

2. **Unit Tests:**
    - Write unit tests to ensure that the `get_storage` method is functioning correctly.
    - Ensure that any potential exceptions and errors, like an invalid address or position, are correctly handled.

3. **Inline Documentation:**
    - Add inline Rust documentation (`///` comments) detailing the purpose and usage of the `get_storage` method.
    - Include any relevant information about the return values and potential errors/exceptions.

### **Task Checklist:**
- [ ] Implement the `get_storage` method.
- [ ] Write unit tests for the `get_storage` method.
- [ ] Add inline documentation for the `get_storage` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### PR DESCRIPTION
**Pull request by nbaztec**
# What :computer: 
* Adds `eth_getStorageAt`
* Allows saving historical block states (currently unbounded)
* Adds `testing::deploy_contract` helper

# Why :hand:
* Allows users to retrieve raw storage items at a specific height, enabling better debugging.

# Evidence :camera:
image

# Notes :memo:
* Fixes #45

### CODE REVIEW

File: src/node.rs [hunk 1]
```diff
@@ -270,6 +273,9 @@ pub struct InMemoryNodeInner<S> {
     pub console_log_handler: ConsoleLogHandler,
     pub system_contracts: SystemContracts,
     pub impersonated_accounts: HashSet<Address>,
+    /// Keeps track of historical states indexed via block hash.
+    /// This is currently unbounded and may change in future to keep only the last N blocks.
+    pub previous_states: HashMap<H256, HashMap<StorageKey, StorageValue>>,
```
Comment (nbaztec) [inline src/node.rs:278]: Unsure if we should already do it or not care about it at all?

File: src/node.rs [hunk 2]
```diff
@@ -2260,13 +2294,94 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EthNamespaceT for
...
+                    .get(&block_number.as_u64())
+                    .and_then(|block_hash| reader.previous_states.get(block_hash))
+                    .and_then(|state| state.get(&storage_key))
+                    .cloned()
+                    .ok_or_else(|| {
+                        log::error!("unable to get storage for block number {}", block_number);
+                        return into_jsrpc_error(Web3Error::InternalError);
+                    })
+            } else {
+                println!("read me {block:?}");
```
Comment (dutterbutter) [inline src/node.rs:2365]:
```suggestion
```

File: src/node.rs (hunk 1 above)
Comment (MexicanAce) [reply src/node.rs:278]: I'm guessing you're referring to the bounding to N blocks? Yeah, this one does slightly worry me for size as we'll be cloning the raw_storage every time. If someone keeps this node up for too long, this HashMap can grow exponentially.

I'm fine letting this go as-is, and just leave this comment to watch out for future space/performance issues.

File: src/testing.rs [hunk 3]
```diff
@@ -442,6 +522,22 @@ impl LogBuilder {
     }
 }
 
+/// Simple storage solidity contract that stores and retrieves two numbers
+///
+/// contract Storage {
+///     uint256 number1 = 1024;
+///     uint256 number2 = 115792089237316195423570985008687907853269984665640564039457584007913129639935;
```
Comment (MexicanAce) [inline src/testing.rs:529]: nit: Might be worth mentioning that this is `U256::MAX` since we use that value for our tests?

**Review by MexicanAce — APPROVED** [review]
LGTM

File: src/node.rs (hunk 1 above)
Comment (grw-ms) [reply src/node.rs:278]: in foundry they keep a .json cache of StorageMap per block- maybe it's worth copying this pattern and using a LRU to keep some bounded number of blocks in memory?

Comment (nbaztec) [reply src/node.rs:278]: The question would be to how long (if at all) to persist it given that it's a test node. We could remove all previous state on start with a disk backed LRU. Just wondering if it's needed and if yes, to what degree since ideally the state shouldn't be huge. Perhaps a follow up ticket to revisit this.

Comment (MexicanAce) [reply src/node.rs:278]: Agreed that if we persist all state on disk, then we'd probably need to remove all files on restart of the node. Perhaps we only cache storage requests to upstream for forks?

I noticed that we're not cache-ing within the `get_storage_at` function of `HttpForkSource`

Comment (nbaztec) [reply src/node.rs:278]: Yeah that's cause initially I decided it would be hard to cache against `latest` block. I think it might be even hard to resolve the latest block number at the `HttpForkSource` given it doesn't know of the node.

Comment (MexicanAce) [reply src/node.rs:278]: Hmmm yeah :/ ... Kk, it seems like the path forward for now is to leave as is and revisit if problems arise. For what it's worth, I pulled down this branch and ran the e2e-tests dozens of times with no impact in response time from the node.  I'm sure Memory usage went up, but I'm not overly worried about that at the moment.

We can also discuss offline tomorrow if that helps

**Review by MexicanAce — DISMISSED** [review]

File: src/node.rs [hunk 4]
```diff
@@ -633,6 +641,27 @@ impl<S: std::fmt::Debug + ForkSource> InMemoryNodeInner<S> {
     pub fn stop_impersonating_account(&mut self, address: Address) -> bool {
         self.impersonated_accounts.remove(&address)
     }
+
+    /// Archives the current state for later queries.
+    pub fn archive_state(&mut self) -> Result<(), String> {
+        if self.previous_states.len() == MAX_PREVIOUS_STATES as usize {
```
Comment (mm-zk) [inline src/node.rs:647]: (let's make it a little bit more defensive - in case someone adds multiple entries to previous_states)

while self.previous_states.len() > MAX_PREVIOUS_STATES

File: src/node.rs [hunk 5]
```diff
@@ -633,6 +641,27 @@ impl<S: std::fmt::Debug + ForkSource> InMemoryNodeInner<S> {
...
+            if let Some(entry) = self.previous_states.shift_remove_index(0) {
+                log::info!("removing archived state for previous block {:#x}", entry.0);
```
Comment (mm-zk) [inline src/node.rs:649]: consider switching to debug (as this will get spammy - will print after every block after 128 blocks)

File: src/node.rs [hunk 6]
```diff
@@ -2284,13 +2324,93 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EthNamespaceT for
...
+                .unwrap_or_else(|| Ok(U64::from(reader.current_miniblock)))?;
+
+            if block_number.as_u64() == reader.current_miniblock {
+                Ok(H256(reader.fork_storage.read_value(&storage_key).0))
+            } else if reader.block_hashes.contains_key(&block_number.as_u64()) {
+                reader
+                    .block_hashes
+                    .get(&block_number.as_u64())
+                    .and_then(|block_hash| reader.previous_states.get(block_hash))
+                    .and_then(|state| state.get(&storage_key))
```
Comment (mm-zk) [inline src/node.rs:2388]: I see two potential bugs here:

Consider this scenario: we forked testnet at block 1000, and did a transaction in block 1001 that did:
* insert(0x1, 0x1)
* delete(0x2)


Now the bugs:
* if we query for key (0x3) at block 1001 - we'll get empty value (as this key is not present in state - as we didn't modify it) - but in practice, we should query the testnet at block 1000 (! - important - at fork level) to get the current value of the key there

* the other bug is about deletions - and it might exist in the current system too (but we'll have to check). When we query at block 1001 for key 0x2 (that we just deleted) - we should return 0x0 then (as the value was deleted) - but from what I see, we'll query the testnet and might return whatever value was set there before..

let's make sure that we have test coverage for these corner cases.

**Review by mm-zk — CHANGES_REQUESTED** [review]

File: src/node.rs (hunk 6 above)
Comment (nbaztec) [reply src/node.rs:2388]: I was under the assumption that forking at block 1000 would copy all the state of the block `1000`. So when the state would be archived/cloned it would still persist the keys that weren't modified. If that's not the case then we could effectively query the fork if the state entry isn't found. Any blocks created later would always have the new state additions in the cloned states.

As for the deletion, I currently see the `set_value` function in `StorageView` which leads me to believe (correct me if I'm wrong) that a deletion is achieved via setting the storage to `H256::zero()` so if a state entry `0x2` is removed via setting it to `H256::zero()` that would still be tracked, isn't it?

File: src/node.rs [hunk 7]
```diff
@@ -646,6 +654,31 @@ impl<S: std::fmt::Debug + ForkSource> InMemoryNodeInner<S> {
...
     }
+
+    /// Archives the current state for later queries.
+    pub fn archive_state(&mut self) -> Result<(), String> {
+        if self.previous_states.len() > MAX_PREVIOUS_STATES as usize {
+            if let Some(entry) = self.previous_states.shift_remove_index(0) {
+                log::debug!("removing archived state for previous block {:#x}", entry.0);
+            }
+        }
+        println!(
```
Comment (dutterbutter) [inline src/node.rs:665]: Nit: use logger please

Comment (nbaztec) [reply src/node.rs]: That wasn't supposed to be there, but I guess it's nice to have in debug log.

File: src/node.rs (hunk 6 above)
Comment (MexicanAce) [reply src/node.rs:2463]: I don't believe we're copying the full state of the block because in `from_url_and_miniblock_and_chain`, we're only fetching `get_block_details` and `get_block_by_hash`. The only information we're saving from the fork on start-up is:
```rust
        ForkDetails {
            fork_source: HttpForkSource::new(url.to_owned(), cache_config),
            l1_block: l1_batch_number,
            l2_block: block,
            block_timestamp: block_details.base.timestamp,
            l2_miniblock: miniblock,
            l2_miniblock_hash: root_hash,
            overwrite_chain_id: chain_id,
            l1_gas_price: block_details.base.l1_gas_price,
        }
```

I can dig into the deletion logic

Comment (nbaztec) [reply src/node.rs:2463]: Yeah you're right on the copying block state. Upon further discovery I realized the mistake of my presumption and added the fix. Still looking into deletion.

File: Cargo.toml [hunk 8]
```diff
@@ -51,8 +51,12 @@ itertools = "0.10.5"
...
+indexmap = "2.0.1"
 
 [dev-dependencies]
 httptest = "0.15.4"
 tempdir = "0.3.7"
-maplit = "1.0.2"
\ No newline at end of file
+maplit = "1.0.2"
+zksync-web3-rs = "0.1.1"
+ethers-contract = "2.0.7"
```
Comment (georgewhewell) [inline Cargo.toml:61]: nit: iirc `ethers_contract` is re-exported by `ethers` at `ethers::contract`- you don't need a separate dependency for it

File: src/node.rs (hunk 1 above)
Comment (MexicanAce) [reply src/node.rs]: PR updated with 128 block FIFO queue of storage

File: src/node.rs (hunk 6 above)
Comment (nbaztec) [reply src/node.rs:2463]: @mm-zk do you have some additional context on the delete? Looking into `zksync-era` repo I was unable to find any references where storage is deleted. As far as solidity smart contracts are concerned, `delete` merely [assigns](https://docs.soliditylang.org/en/develop/types.html#delete) the zero-value of the type to a key. Thus the deleted key must be present in `modified_keys` and subsequently set to 0.

**Review by MexicanAce — APPROVED** [review]
LGTM
